shared/js/background/onboarding.es6.js

```diff
@@ -0,0 +1,117 @@
+            if (e.origin === window.location.origin && e.data.type === 'documentStartDataResponse') {
+                window.removeEventListener('message', handleMessage)
+                cb(null, e.data.payload)
+            }
+        })
+    }
+
+    function start () {
+        // Consolidate data: grab value of the Chrome Dialogue content-script
+        // this script was injected earlier to capture value as early as possible
+        getDocumentStartData((err, documentStartData) => {
+            if (err) {
+                console.error(err)
+            }
+
+            // DDG privacy policy prevents us to use `chrome.runtime` on the SERP so we
+            // setup a relay here so that the SERP can assess the background process
+            if (browser === 'chrome') {
+                window.addEventListener('message', (e) => {
+                    if (e.origin === window.location.origin) {
+                        switch (e.data.type) {
+                        case 'healthCheckRequest': {
+                            try {
+                                chrome.runtime.sendMessage(extensionId, e.data.type, (response) => {
+                                    e.source.postMessage(
+                                        { type: 'healthCheckResponse', isAlive: !chrome.runtime.lastError },
+                                        e.origin
+                                    )
+                                })
+                            } catch (err) {
+                                e.source.postMessage(
+                                    { type: 'healthCheckResponse', isAlive: false },
+                                    e.origin
+                                )
+                            }
+                            break
+                        }
+
+                        case 'rescheduleCounterMessagingRequest': {
+                            chrome.runtime.sendMessage(extensionId, e.data.type, (response) => {
+                                if (chrome.runtime.lastError) {
+                                    console.error(chrome.runtime.lastError)
+                                }
+                            })
+                            break
+                        }
+                        }
+                    }
+                })
+            }
+
+            const args = Object.assign({
+                isAddressBarQuery,
+                showWelcomeBanner,
+                showCounterMessaging
+            }, documentStartData)
+
+            // the content script do not share the same `window` as the page
+            // so we inject a `<script>` to be able to access the page `window`
+            const script = document.createElement('script')
+            script.textContent = `
+                    if (window.onFirstSearchPostExtensionInstall) {
+                        window.onFirstSearchPostExtensionInstall(${JSON.stringify(args)})
```

Comment:
Can this not be done with post messaging also?

Reply:
So my understanding is that we wanted to avoid adding an event listener to the SERP (we would need that event listener for all queries which was considered wasteful).
I fully agree that the current approach is gross but I think that it's the only way to avoid an event listener on the SERP and make sure that the SERP is only impacted on the first search post extension install.

Reply:
Thanks for the explanation!